api/include/opentelemetry/trace/span_context_kv_iterable_view.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+#include "opentelemetry/common/key_value_iterable_view.h"
+#include "opentelemetry/nostd/utility.h"
+#include "opentelemetry/trace/span_context_kv_iterable.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace trace
+{
+namespace detail
+{
+template <class T>
+inline void take_span_context_kv(SpanContext, common::KeyValueIterableView<T>)
+{}
+
+template <class T, nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
+inline void take_span_context_kv(SpanContext, T &)
+{}
+
+inline void take_span_context_kv(
+    SpanContext,
+    std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>>)
+{}
+
+template <class T>
+auto is_span_context_kv_iterable_impl(T iterable)
+    -> decltype(take_span_context_kv(std::begin(iterable)->first, std::begin(iterable)->second),
+                nostd::size(iterable),
+                std::true_type{});
+
+std::false_type is_span_context_kv_iterable_impl(...);
+
+template <class T>
+struct is_span_context_kv_iterable
+{
+  static const bool value =
+      decltype(detail::is_span_context_kv_iterable_impl(std::declval<T>()))::value;
+};
+}  // namespace detail
+
+template <class T>
+class SpanContextKeyValueIterableView final : public SpanContextKeyValueIterable
+{
+  static_assert(detail::is_span_context_kv_iterable<T>::value,
+                "Must be a context/key-value iterable");
+
+public:
+  explicit SpanContextKeyValueIterableView(const T &links) noexcept : container_{&links} {}
+
+  bool ForEachKeyValue(
+      nostd::function_ref<bool(SpanContext, opentelemetry::common::KeyValueIterable &)> callback)
+      const noexcept override
```

Comment:
```suggestion
  bool ForEachKeyValue(
      nostd::function_ref<bool(SpanContext, const opentelemetry::common::KeyValueIterable &)> callback)
      const noexcept override
```

I think we can use a `const` here, this should save you the `const_cast` below. All methods that `KeyValueIterable` provides are `const` anyway.

Reply:
Good point. This needed changes in other places too, fixed now.